### DISCUSSION TITLE
How do I get worker kv working locally without internet in dev mode?

### DISCUSSION THREAD

**Discussion by fivesheep** [cloudflare/workers-sdk#5573] [Q&A]
Say, I want to do integration tests in our ci env which internet access is limited. my worker has dependency on worker kv, but i don't want it to connect to any remote place given the limit.

I had tried the following in the toml file:

```
[env.ci]
name = "my-local-worker"
send_metrics = false
workers_dev = true
kv_namespaces = [
  { binding = "SOME_THING", id = "some-fake-id"},
]
``` 
and started with this command `npx wrangler dev -e ci`  got the following errors

```
✘ [ERROR] In development, you should use a separate kv namespace than the one you'd use in production. Please create a new kv namespace with "wrangler kv:namespace create <name> --preview" and add its id as preview_id to the kv_namespace "SOME_THING" in your wrangler.toml
```
I can't run this command because no internet access, and it also asked for opening an url with browser, which is not possible in an automated ci env.

from the official doc, it mentioned
> Resources such as KV, Durable Objects, D1, and R2 will be stored and persisted locally and not affect live production or preview data. Wrangler will automatically create local versions of bindings found in wrangler.toml. These will not have data in them initially, so you will need to add data manually.

so what does it really mean by `Wrangler will automatically create local versions of bindings found in wrangler.toml.` ?  how can I setup test this from my ci env?

**Comment by mootari** [comment]
Are you using wrangler v2 or v3? v3 is local by default, and as far as I can tell you shouldn't see the warning unless you're running the worker remotely:
https://github.com/cloudflare/workers-sdk/blob/1b80decfaf56c8782e49dad685c344288629b668/packages/wrangler/src/dev.ts#L892-L898

**Comment by flikteoh** [comment]
This error means that you need to add a preview_id in your wrangler.toml for the env kv you’re trying to use locally. It can be the same or you could create a new kv, which wrangler will create in .wrangler/ folder locally.
